src/Components/Components/src/Lifetime/ComponentApplicationState.cs

```diff
@@ -0,0 +1,152 @@
+        {
+            if (_existingState != null)
+            {
+                throw new InvalidOperationException("ComponentApplicationState already initialized.");
+            }
+            _existingState = existingState ?? throw new ArgumentNullException(nameof(existingState));
+        }
+
+        /// <summary>
+        /// Registers a callback that will be invoked when the component application is being paused.
+        /// </summary>
+        /// <param name="callback">The <see cref="Func{TResult}"/> to invoke.</param>
+        public void RegisterOnPersistingCallback(Func<Task> callback)
+        {
+            if (callback is null)
+            {
+                throw new ArgumentNullException(nameof(callback));
+            }
+
+            _registeredCallbacks.Add(callback);
+        }
+
+        /// <summary>
+        /// Tries to retrieve the persisted state with the given <paramref name="key"/>.
+        /// </summary>
+        /// <param name="key">The key used to persist the state.</param>
+        /// <param name="value">The persisted state.</param>
+        /// <returns><c>true</c> if the state was found;<c>false</c> otherwise.</returns>
+        public bool TryRetrievePersistedState(string key, [MaybeNullWhen(false)] out byte[] value)
+        {
+            if (key is null)
+            {
+                throw new ArgumentNullException(nameof(key));
+            }
+
+            if (_existingState == null)
+            {
+                // Services during prerendering might try to access their state upon injection on the page
+                // and we don't want to fail in that case.
+                // When a service is prerendering there is no state to restore and in other cases the host
+                // is responsible for initializing the state before services or components can access it.
+                value = null;
+                return false;
+            }
+
+            if (_existingState.TryGetValue(key, out value))
+            {
+                _existingState.Remove(key);
+                return true;
+            }
+            else
+            {
+                return false;
+            }
+        }
+
+        /// <summary>
+        /// Persists the serialized state <paramref name="value"/> for the given <paramref name="key"/>.
+        /// </summary>
+        /// <param name="key">The key to use to persist the state.</param>
+        /// <param name="value">The state to persist.</param>
+        public void PersistState(string key, byte[] value)
+        {
+            if (key is null)
+            {
+                throw new ArgumentNullException(nameof(key));
+            }
+
+            if (value is null)
+            {
+                throw new ArgumentNullException(nameof(value));
+            }
+
+            if (_currentState.ContainsKey(key))
+            {
+                throw new ArgumentException($"There is already a persisted object under the same key '{key}'");
```

Comment:
This also seems pretty aggressive. What if there might be multiple instances of a state-preserving component? The developer has to invent a system to give them unique keys.

Possibilities:

 * Change this to `TryPreserveState`, using the same semantics as `Dictionary.TryAdd`
 * Or at least offer a `bool HasState(key)` method so the developer can avoid causing an exception, though that's not as convenient

Reply:
State should be private to each component/service, you need to account for that when you are designing the state preservation of your component, if you need to share, you can put the data in a common service.

Reply:
OK, makes sense. Thanks.